File: C++/anagram_strings.cpp
class Solution {
public:
    bool compare(int a[], int b[], int n) {
        for (int i = 0; i < n; ++i) {
            if (a[i] != b[i]) return false;
        }
        
        return true;
    }
    
    // Time complexity: O(n)
    // Space complexity: O(k)
    vector<int> findAnagrams(string s, string p) {
        int ks[26] {0};
        int kp[26] {0};
        vector<int> ret;
        int ls = s.size();
        int lp = p.size();
        if (ls < lp) return {};
        
        for (int i = 0; i < lp; ++i) {
            ks[s[i] - 'a'] += 1;
            kp[p[i] - 'a'] += 1;
        }
        
        if (compare(ks, kp, 26)) ret.push_back(0);
        for (int i = 1; i < ls - lp + 1; ++i) {
            ks[s[i-1] - 'a'] -= 1;
            ks[s[i+lp-1] - 'a'] += 1;
            if (compare(ks, kp, 26)) {
                ret.push_back(i);
            }
        }
        
        return ret;
    }
};

File: Python/anagram_strings.py
class Solution(object):
    def __int__(self):
        self.name = 'S->13'

    # Time complexity: O(n)
    # Space complexity: O(k)
    def anagramsString(self, s, p):
        ret = []
        ks = [0] * 26
        kp = [0] * 26
        lp = len(p)
        ll = len(s)
        if ll < lp:
            return []
        for i in range(lp):
            ks[ord(s[i]) - ord('a')] += 1
            kp[ord(p[i]) - ord('a')] += 1
        if (kp == ks): ret.append(0)

        for i in range(1, ll - lp + 1):
            ks[ord(s[i - 1]) - ord('a')] -= 1
            ks[ord(s[i + lp - 1]) - ord('a')] += 1
            if ks == kp:
                ret.append(i)
        return ret
s = 'cbaebabacb'
p = 'abc'

print(Solution().anagramsString(s, p))
